sapicore/engine/ensemble/__init__.py

```diff
@@ -0,0 +1,66 @@
+""" Ensembles (layers) manage neuron collections by fixing their size, extending attribute tensors,
+and handling layer-aware operations such as heterogeneous parameter initialization
+and normalization.
+
+Note
+----
+Ensemble instance attributes are 1D tensors corresponding to potentially heterogeneous parameters
+and states of neurons (tensor elements) whose behavior is prescribed by their parent class
+(e.g., :class:`~engine.neuron.spiking.LIF.LIFNeuron`).
+
+Ensembles can be initialized programmatically or from a configuration YAML containing neuron parameters and their
+generation method (fixed, zipped, grid, or drawn from a distribution). Forward calls invoke the parent neuron class
+method.
+
+Example
+-------
+Initialize an ensemble of basic analog neurons, inheriting default parameters and behavior from
+:class:`~engine.neuron.analog.AnalogNeuron`:
+
+    >>> from sapicore.engine.ensemble.spiking import LIFEnsemble
+    >>> layer = LIFEnsemble(num_units = 10)
+
+"""
+import torch
+from torch import Tensor
+
+from sapicore.engine.neuron import Neuron
+
+
+class Ensemble(Neuron):
+    """Ensemble base class. Provides generic implementations of tensor expansion and parameter diversification."""
+
+    def __init__(self, num_units: int = 1, **kwargs):
+        """Constructs a generic ensemble instance, inheriting attributes from :class:`~engine.neuron.Neuron`."""
+        super().__init__(**kwargs)
+        self.num_units = num_units
```

Comment:
Aha so this is where `num_units` is defined. But, if synapses are aware of this and base neuron classes are defaulted to have 1 num units, then it's better to set num_units to 1 in the base neuron class as a class variable and doc that it may not be changed (you can do it by making it an `@property` where the setter raises an error). And overwrite it in Ensemble to be allowed to be overwritten. This way it's all consistent.

Reply:
Not as elegant as I'd like it to be, but that's the cost of splitting Neurons and Ensembles :) Implemented your suggestion.